demos/common/processing/async_pipeline.h

```diff
@@ -0,0 +1,150 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#define MOVING_AVERAGE_SAMPLES 5
+
+#pragma once
+#include <string>
+#include <deque>
+#include <map>
+#include <condition_variable>
+#include "config_factory.h"
+#include "requests_pool.h"
+#include "results.h"
+#include "model_base.h"
+
+/// This is base class for asynchronous pipeline
+/// Derived classes should add functions for data submission and output processing
+class PipelineBase
+{
+public:
+    struct PerformanceInfo
+    {
+        int64_t framesCount = 0;
+        std::chrono::steady_clock::duration latencySum = std::chrono::steady_clock::duration::zero();
+        std::chrono::steady_clock::duration lastInferenceLatency = std::chrono::steady_clock::duration::zero();
+        std::chrono::steady_clock::time_point startTime;
+        double movingAverageLatencyMs;
+        uint32_t numRequestsInUse;
+        double movingAverageFPS;
+        double FPS=0;
+
+        double getTotalAverageLatencyMs() const {
+            return ((double)std::chrono::duration_cast<std::chrono::milliseconds>(latencySum).count()) / framesCount;
+        }
+
+        double getLastInferenceLatencyMs() const {
+            return ((double)std::chrono::duration_cast<std::chrono::milliseconds>(lastInferenceLatency).count());
+        }
+
+    };
+
+    struct PerformanceInternalCounters
+    {
+        long long latenciesMs[MOVING_AVERAGE_SAMPLES] = {};
+        std::chrono::steady_clock::time_point retrievalTimestamps[MOVING_AVERAGE_SAMPLES] = {};
+        long long movingLatenciesSumMs = 0;
+        int currentIndex = 0;
+    };
+
+public:
+    /// Loads model and performs required initialization
+    /// @param modelInstance pointer to model object. Object it points to should not be destroyed manually after passing pointer to this function.
+    /// @param cnnConfig - fine tuning configuration for CNN model
+    /// @param engine - pointer to InferenceEngine::Core instance to use.
+    /// If it is omitted, new instance of InferenceEngine::Core will be created inside.
+    PipelineBase(ModelBase* modelInstance, const CnnConfig& cnnConfig, InferenceEngine::Core* engine = nullptr);
+    virtual ~PipelineBase();
+
+    /// Waits until either output data becomes available or pipeline allows to submit more input data.
+    /// @param shouldKeepOrder if true, function will treat results as ready only if next sequential result (frame) is
+    /// ready (so results can be extracted in the same order as they were submitted). Otherwise, function will return if any result is ready.
+    void waitForData(bool shouldKeepOrder = true);
+
+    /// Returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing.
+    /// @returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing, false otherwise.
```

Comment:
This is the same sentence again, it's completely redundant.

Reply:
yes, but I have to keep it to be compatible with doxygen comments structure - general escription and return value are different sections

Reply:
No you don't. We don't even run Doxygen on these sources (or any other sources). And even if we did, well... `@returns` is optional.

Reply:
it's optional, but it's a sign of good style to have it (if function returns something).

Reply:
If you have to repeat the same sentence twice, that cannot be good style.